default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
+    if tech_name in ["SHP_MIL_ROBO_CONT", "SHP_MIDCOMB_LOG"]:
+        return robotic
+    if tech_name in ["SHP_SPACE_FLUX_DRIVE", "SHP_TRANSSPACE_DRIVE"]:
+        return min(robotic, flux)
+    if tech_name in ["SHP_CONTGRAV_MAINT", "SHP_MASSPROP_SPEC"]:
+        return min(robotic, contgrav)
+    if tech_name in ["SHP_NANOROBO_MAINT"]:
+        return min(robotic, nanorobo)
+    if tech_name in ["SHP_XENTRONIUM_HULL"]:
+        return default # this is not a robotic hull!
+
+    # Asteroid hulls
+    if tech_name in ["SHP_ASTEROID_HULLS", "SHP_ASTEROID_REFORM", "SHP_MONOMOLEC_LATTICE", "SHP_SCAT_AST_HULL"]:
+        return asteroid
+    if tech_name in ["SHP_HEAVY_AST_HULL"]:
+        return min(asteroid, astheavy)
+    if tech_name in ["SHP_CAMO_AST_HULL"]:
+        return min(asteroid, astcamo)
+    if tech_name in ["SHP_MINIAST_SWARM"]:
+        return min(asteroid, astswarm)
+
+    # Organic hulls
+    if tech_name in ["SHP_ORG_HULL", "SHP_SENT_HULL"]:
+        return org
+    if tech_name in ["SHP_MULTICELL_CAST", "SHP_ENDOCRINE_SYSTEMS", "SHP_CONT_BIOADAPT"]:
+        return min(org, orgraven)
+    if tech_name in ["SHP_MONOCELL_EXP", "SHP_CONT_SYMB", "SHP_BIOADAPTIVE_SPEC"]:
+        return min(org, orgneural)
+    if tech_name in ["SHP_ENDOSYMB_HULL"]:
+        return min(org, orgendosym)
+
+    # energy hulls
+    if tech_name in ["SHP_FRC_ENRG_COMP"]:
+        return energy
+    if tech_name in ["SHP_QUANT_ENRG_MAG"]:
+        return min(energy, energymag, immediate if bright_star or blackhole else useless)
+    if tech_name in ["SHP_ENRG_BOUND_MAN"]:
+        return min(energy, energyfrac, immediate if bright_star or blackhole else useless)
+    if tech_name in ["SHP_SOLAR_CONT"]:
+        return min(energy, immediate if red_star or blackhole else useless) # red star can be turned into blackhole by Artificial black hole
+
+    # damage control
+    if tech_name in ["SHP_REINFORCED_HULL"]:
+        return default
+    if tech_name in ["SHP_BASIC_DAM_CONT", "SHP_FLEET_REPAIR", "SHP_ADV_DAM_CONT"]:
+        return repair
+
+    # ship parts
+    if tech_name in ["SHP_IMPROVED_ENGINE_COUPLINGS", "SHP_N_DIMENSIONAL_ENGINE_MATRIX", "SHP_SINGULARITY_ENGINE_CORE"]:
+        return engine
+    if tech_name in ["SHP_DEUTERIUM_TANK", "SHP_ANTIMATTER_TANK", "SHP_ZERO_POINT"]:
+        return fuel
+    if tech_name in ["SHP_NOVA_BOMB", "SHP_DEATH_SPORE", "SHP_BIOTERM"]:
+        return useless # AI does not use Nova bomb nor bioterror
+
+    # ship shields
+    if tech_name in ["SHP_DEFLECTOR_SHIELD", "SHP_PLASMA_SHIELD", "SHP_BLACKSHIELD"]:
+        return shield
+    if tech_name in ["SHP_MULTISPEC_SHIELD"]:
+        return stealth
+
+    # ship armor
+    if tech_name in ["SHP_ROOT_ARMOR", "SHP_ZORTRIUM_PLATE"]:
+        return armor
+    if tech_name in ["SHP_DIAMOND_PLATE", "SHP_XENTRONIUM_PLATE"]:
+        return useless if asteroid or neutron else armor # asteroid hull line and neutronium extraction includes better armors
```

Comment:
This seems wrong to me. ```asteroid``` seems to be at offtrack in worst case which is a non-zero priority. Thus ```if asteroid``` will always be True. Instead, you probably intended to check for ```asteroid == hull```.